Bug in sequential repeat when the layer has no parameters

**problems Statement**
In `init` mode, the repeated layer will be reset. In `Pytorch`, we used `reset_parameters` function to reset the parameters of layers, as here:
https://github.com/pfnet/pytorch-pfn-extras/blob/92dad97e3a3eaafb9eb2b2b804266b5ef922e4c7/pytorch_pfn_extras/nn/modules/extended_sequential.py#L15

However, there are layers have no parameters nor `reset_parameters`, such as `torch.nn.ReLU`. An error is raised when the model contains such layer.

**Error Message**
```
pytorch_pfn_extras/nn/modules/extended_sequential.py:68: in repeat
    model_list.append(self._copy_model(mode))
pytorch_pfn_extras/nn/modules/extended_sequential.py:27: in _copy_model
    return _reset_parameters(copy.deepcopy(self))
pytorch_pfn_extras/nn/modules/extended_sequential.py:9: in _reset_parameters
    _reset_parameters(submodel)
pytorch_pfn_extras/nn/modules/extended_sequential.py:17: in _reset_parameters
    model.reset_parameters()
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _

self = ReLU(), name = 'reset_parameters'

    def __getattr__(self, name):
        if '_parameters' in self.__dict__:
            _parameters = self.__dict__['_parameters']
            if name in _parameters:
                return _parameters[name]
        if '_buffers' in self.__dict__:
            _buffers = self.__dict__['_buffers']
            if name in _buffers:
                return _buffers[name]
        if '_modules' in self.__dict__:
            modules = self.__dict__['_modules']
            if name in modules:
                return modules[name]
        raise AttributeError("'{}' object has no attribute '{}'".format(
>           type(self).__name__, name))
E       AttributeError: 'ReLU' object has no attribute 'reset_parameters'
```

--- a/pytorch_pfn_extras/nn/modules/extended_sequential.py
+++ b/pytorch_pfn_extras/nn/modules/extended_sequential.py
@@ -17,9 +17,7 @@ def _reset_parameters(model):
         elif hasattr(model, '_reset_parameters'):
             model._reset_parameters()
         else:
-            module_name = model.__module__
-            if not (module_name.startswith('torch.nn') or
-                    module_name.startswith('pytorch_pfn_extras.nn')):
+            if len(list(model.parameters())) != 0:
                 warnings.warn('Cannot reset the parameters of module {}. '
                               'Consider adding `reset_parameters` or'
                               '`_reset_parameters`'
@@ -52,7 +50,7 @@ class ExtendedSequential(torch.nn.Sequential):
         The functions is supposed to behave the same way as `repeat`
         in `chainer`.
 
-        For user-defined module, add a ``reset_parameters`` or
+        For user-defined module with parameters, add a ``reset_parameters`` or
         ``_reset_parameters`` function to repeat with mode ``init``.
         Otherwise, a warning message is generated.
 

--- a/pytorch_pfn_extras/nn/modules/extended_sequential.py
+++ b/pytorch_pfn_extras/nn/modules/extended_sequential.py
@@ -15,9 +15,6 @@ def _reset_parameters(model):
             model.reset_parameters()
         elif hasattr(model, '_reset_parameters'):
             model._reset_parameters()
-        else:
-            raise RuntimeError(
-                "Can not reset parameters of model {}".format(model))
     return model
 
 

--- a/pytorch_pfn_extras/nn/modules/extended_sequential.py
+++ b/pytorch_pfn_extras/nn/modules/extended_sequential.py
@@ -11,7 +11,8 @@ def _reset_parameters(model):
         for submodel in model.values():
             _reset_parameters(submodel)
     else:
-        if isinstance(model, torch.nn.Module):
+        if (isinstance(model, torch.nn.Module) and
+           hasattr(model, 'reset_parameters')):
             model.reset_parameters()
     return model
 

--- a/pytorch_pfn_extras/nn/modules/extended_sequential.py
+++ b/pytorch_pfn_extras/nn/modules/extended_sequential.py
@@ -11,9 +11,13 @@ def _reset_parameters(model):
         for submodel in model.values():
             _reset_parameters(submodel)
     else:
-        if (isinstance(model, torch.nn.Module) and
-           hasattr(model, 'reset_parameters')):
+        if hasattr(model, 'reset_parameters'):
             model.reset_parameters()
+        elif hasattr(model, '_reset_parameters'):
+            model._reset_parameters()
+        else:
+            raise RuntimeError(
+                "Can not reset parameters of model {}".format(model))
     return model
 
 

--- a/pytorch_pfn_extras/nn/modules/extended_sequential.py
+++ b/pytorch_pfn_extras/nn/modules/extended_sequential.py
@@ -52,6 +52,10 @@ class ExtendedSequential(torch.nn.Sequential):
         The functions is supposed to behave the same way as `repeat`
         in `chainer`.
 
+        For user-defined module, add a ``reset_parameters`` or
+        ``_reset_parameters`` function to repeat with mode ``init``.
+        Otherwise, a warning message is generated.
+
         Args:
             n_repeat (int): Number of times to repeat.
             mode (str): It should be either ``init``, ``copy``, or ``share``.

--- a/pytorch_pfn_extras/nn/modules/extended_sequential.py
+++ b/pytorch_pfn_extras/nn/modules/extended_sequential.py
@@ -20,8 +20,8 @@ def _reset_parameters(model):
             if (len(list(model.parameters())) != 0 or
                     len(list(model.buffers())) != 0):
                 warnings.warn('Cannot reset the parameters of module {}. '
-                              'Consider adding `reset_parameters` or'
-                              '`_reset_parameters`'
+                              'Consider adding `reset_parameters` or '
+                              '`_reset_parameters` '
                               'functions to the module'.format(model),
                               UserWarning)
 

--- a/pytorch_pfn_extras/nn/modules/extended_sequential.py
+++ b/pytorch_pfn_extras/nn/modules/extended_sequential.py
@@ -1,5 +1,6 @@
 import torch
 import copy
+import warnings
 
 
 def _reset_parameters(model):
@@ -15,6 +16,16 @@ def _reset_parameters(model):
             model.reset_parameters()
         elif hasattr(model, '_reset_parameters'):
             model._reset_parameters()
+        else:
+            module_name = model.__module__
+            if not (module_name.startswith('torch.nn') or
+                    module_name.startswith('pytorch_pfn_extras.nn')):
+                warnings.warn('Cannot reset the parameters of module {}. '
+                              'Consider adding `reset_parameters` or'
+                              '`_reset_parameters`'
+                              'functions to the module'.format(model),
+                              UserWarning)
+
     return model
 
 

--- a/pytorch_pfn_extras/nn/modules/extended_sequential.py
+++ b/pytorch_pfn_extras/nn/modules/extended_sequential.py
@@ -51,9 +51,17 @@ class ExtendedSequential(torch.nn.Sequential):
         The functions is supposed to behave the same way as `repeat`
         in `chainer`.
 
-        For user-defined module with parameters, add a ``reset_parameters`` or
-        ``_reset_parameters`` function to repeat with mode ``init``.
-        Otherwise, a warning message is generated.
+        When the mode is set to ``init``, the default value,
+        modules will be copied and reinitialized by calling
+        ``reset_parameters`` (or ``_reset_parameters``) method.
+
+        To repeat user-defined modules, which have parameters or buffers,
+        with mode=``init`` in this Sequential,
+        you need to implement the ``reset_parameters`` or ``_reset_parameters``
+        method to the module to reinitialize parameters
+        and (if necessary) buffers;
+        otherwise the initialization cannot be performed
+        and a warning message will be shown.
 
         Args:
             n_repeat (int): Number of times to repeat.

--- a/pytorch_pfn_extras/nn/modules/extended_sequential.py
+++ b/pytorch_pfn_extras/nn/modules/extended_sequential.py
@@ -17,7 +17,8 @@ def _reset_parameters(model):
         elif hasattr(model, '_reset_parameters'):
             model._reset_parameters()
         else:
-            if len(list(model.parameters())) != 0:
+            if (len(list(model.parameters())) != 0 or
+                    len(list(model.buffers())) != 0):
                 warnings.warn('Cannot reset the parameters of module {}. '
                               'Consider adding `reset_parameters` or'
                               '`_reset_parameters`'

--- a/tests/pytorch_pfn_extras_tests/nn_tests/modules_tests/test_extended_sequential.py
+++ b/tests/pytorch_pfn_extras_tests/nn_tests/modules_tests/test_extended_sequential.py
@@ -18,6 +18,8 @@ class TestExtendedSequential(object):
         self.l1 = ppe.nn.LazyLinear(None, 3)
         self.l2 = nn.Linear(3, 2)
         self.l3 = nn.Linear(2, 3)
+        # a layer without reset_parameters
+        self.l4 = nn.ReLU()
         # s1: l1 -> l2
         if module == nn.Sequential:
             self.s1 = module(self.l1, self.l2)
@@ -28,11 +30,11 @@ class TestExtendedSequential(object):
         else:
             self.s1 = module([self.l1, self.l2])
         self.module = module
-        # s2: s1 (l1 -> l2) -> l3
-        self.s2 = ppe.nn.ExtendedSequential(self.s1, self.l3)
+        # s2: s1 (l1 -> l2) -> l3 -> l4
+        self.s2 = ppe.nn.ExtendedSequential(self.s1, self.l3, self.l4)
 
     def test_repeat_with_init(self):
-        # s2 ((l1 -> l2) -> l3) -> s2 ((l1 -> l2) -> l3)
+        # s2 ((l1 -> l2) -> l3 -> l4) -> s2 ((l1 -> l2) -> l3 -> l4)
         ret = self.s2.repeat(2)
         assertions.assertIsNot(ret[0], self.s2)
         assertions.assertIs(type(ret[0]), type(self.s2))
@@ -70,7 +72,7 @@ class TestExtendedSequential(object):
         assertions.assertEqual(len(ret), 0)
 
     def test_repeat_with_copy(self):
-        # s2 ((l1 -> l2) -> l3) -> s2 ((l1 -> l2) -> l3)
+        # s2 ((l1 -> l2) -> l3 -> l4) -> s2 ((l1 -> l2) -> l3 -> l4)
         ret = self.s2.repeat(2, mode='copy')
         assertions.assertIsNot(ret[0], self.s2)
         assertions.assertIs(type(ret[0]), type(self.s2))
@@ -106,7 +108,7 @@ class TestExtendedSequential(object):
         assertions.assertEqual(len(ret), 0)
 
     def test_repeat_with_share(self):
-        # s2 ((l1 -> l2) -> l3) -> s2 ((l1 -> l2) -> l3)
+        # s2 ((l1 -> l2) -> l3 -> l4) -> s2 ((l1 -> l2) -> l3 -> l4)
         ret = self.s2.repeat(2, mode='share')
         assertions.assertIsNot(ret[0], self.s2)
         assertions.assertIs(type(ret[0]), type(self.s2))

--- a/tests/pytorch_pfn_extras_tests/nn_tests/modules_tests/test_extended_sequential.py
+++ b/tests/pytorch_pfn_extras_tests/nn_tests/modules_tests/test_extended_sequential.py
@@ -216,7 +216,7 @@ def test_no_warning_when_repeat(module):
 
 
 @pytest.mark.parametrize('module', [
-    UserDefinedLayerWithParameters
+    UserDefinedLayerWithParameters,
 ])
 def test_warning_when_repeat(module):
     model = ppe.nn.ExtendedSequential(module())

--- a/tests/pytorch_pfn_extras_tests/nn_tests/modules_tests/test_extended_sequential.py
+++ b/tests/pytorch_pfn_extras_tests/nn_tests/modules_tests/test_extended_sequential.py
@@ -3,6 +3,7 @@ import pytest
 import functools
 
 import numpy
+import torch
 from torch import nn
 import pytorch_pfn_extras as ppe
 
@@ -184,9 +185,26 @@ class UserDefinedLayerWithUnderScoreReset(nn.Module):
         pass
 
 
-@pytest.mark.parametrize('module', [nn.ReLU,
-                                    UserDefinedLayerWithUnderScoreReset,
-                                    UserDefinedLayerWithReset])
+class UserDefinedLayerWithParameters(nn.Module):
+    def __init__(self):
+        super().__init__()
+        param = nn.Parameter(torch.zeros(1, 1))
+        self.register_parameter('weight', param)
+
+    def forward(self):
+        pass
+
+
+@pytest.mark.parametrize('module', [
+    # buit-in, no parameters
+    nn.ReLU,
+    # no parameters
+    UserDefinedLayer,
+    # has `_reset_parameters`
+    UserDefinedLayerWithUnderScoreReset,
+    # has `reset_parameters`
+    UserDefinedLayerWithReset,
+])
 def test_no_warning_when_repeat(module):
     model = ppe.nn.ExtendedSequential(module())
     # no warnings are raised on these modules
@@ -194,7 +212,9 @@ def test_no_warning_when_repeat(module):
         model.repeat(2)
 
 
-@pytest.mark.parametrize('module', [UserDefinedLayer])
+@pytest.mark.parametrize('module', [
+    UserDefinedLayerWithParameters
+])
 def test_warning_when_repeat(module):
     model = ppe.nn.ExtendedSequential(module())
     # warnings are raised on these modules

--- a/tests/pytorch_pfn_extras_tests/nn_tests/modules_tests/test_extended_sequential.py
+++ b/tests/pytorch_pfn_extras_tests/nn_tests/modules_tests/test_extended_sequential.py
@@ -160,3 +160,43 @@ class TestExtendedSequential(object):
         assertions.assertEqual(len(ret), 2)
         ret = self.s2.repeat(0, mode='share')
         assertions.assertEqual(len(ret), 0)
+
+
+class UserDefinedLayerWithReset(nn.Module):
+    def __init__(self):
+        super().__init__()
+
+    def forward(self):
+        pass
+
+    def reset_parameters(self):
+        pass
+
+
+class UserDefinedLayerWithUnderScoreReset(nn.Module):
+    def __init__(self):
+        super().__init__()
+
+    def forward(self):
+        pass
+
+    def _reset_parameters(self):
+        pass
+
+
+@pytest.mark.parametrize('module', [nn.ReLU,
+                                    UserDefinedLayerWithUnderScoreReset,
+                                    UserDefinedLayerWithReset])
+def test_no_warning_when_repeat(module):
+    model = ppe.nn.ExtendedSequential(module())
+    # no warnings are raised on these modules
+    with pytest.warns(None):
+        model.repeat(2)
+
+
+@pytest.mark.parametrize('module', [UserDefinedLayer])
+def test_warning_when_repeat(module):
+    model = ppe.nn.ExtendedSequential(module())
+    # warnings are raised on these modules
+    with pytest.warns(UserWarning):
+        model.repeat(2)

--- a/tests/pytorch_pfn_extras_tests/nn_tests/modules_tests/test_extended_sequential.py
+++ b/tests/pytorch_pfn_extras_tests/nn_tests/modules_tests/test_extended_sequential.py
@@ -1,5 +1,6 @@
 import unittest
 import pytest
+import functools
 
 import numpy
 from torch import nn
@@ -11,15 +12,27 @@ assertions = unittest.TestCase('__init__')
 @pytest.mark.parametrize('module', [nn.Sequential,
                                     nn.ModuleList,
                                     nn.ModuleDict])
+@pytest.mark.parametrize('irregular_layer', [
+    # No reset_parameters
+    nn.ReLU,
+    # use reset_running_stats
+    functools.partial(
+        nn.BatchNorm1d, 1),
+    # use _reset_parameters
+    functools.partial(
+        nn.MultiheadAttention, 1, 1),
+    # ppe.nn layer
+    functools.partial(
+        ppe.nn.LazyConv1d, None, 1, 1)])
 class TestExtendedSequential(object):
 
     @pytest.fixture(autouse=True)
-    def setUp(self, module):
+    def setUp(self, module, irregular_layer):
         self.l1 = ppe.nn.LazyLinear(None, 3)
         self.l2 = nn.Linear(3, 2)
         self.l3 = nn.Linear(2, 3)
         # a layer without reset_parameters
-        self.l4 = nn.ReLU()
+        self.l4 = irregular_layer()
         # s1: l1 -> l2
         if module == nn.Sequential:
             self.s1 = module(self.l1, self.l2)

--- a/tests/pytorch_pfn_extras_tests/nn_tests/modules_tests/test_extended_sequential.py
+++ b/tests/pytorch_pfn_extras_tests/nn_tests/modules_tests/test_extended_sequential.py
@@ -9,10 +9,21 @@ import pytorch_pfn_extras as ppe
 assertions = unittest.TestCase('__init__')
 
 
+class UserDefinedLayer(nn.Module):
+    def __init__(self):
+        super().__init__()
+
+    def forward(self):
+        pass
+
+    def reset_parameters(self):
+        pass
+
 @pytest.mark.parametrize('module', [nn.Sequential,
                                     nn.ModuleList,
                                     nn.ModuleDict])
 @pytest.mark.parametrize('irregular_layer', [
+    UserDefinedLayer,
     # No reset_parameters
     nn.ReLU,
     # use reset_running_stats

--- a/tests/pytorch_pfn_extras_tests/nn_tests/modules_tests/test_extended_sequential.py
+++ b/tests/pytorch_pfn_extras_tests/nn_tests/modules_tests/test_extended_sequential.py
@@ -198,6 +198,15 @@ class UserDefinedLayerWithParameters(nn.Module):
         pass
 
 
+class UserDefinedLayerWithBuffer(nn.Module):
+    def __init__(self):
+        super().__init__()
+        self.register_buffer('weight', torch.zeros(1, 1))
+
+    def forward(self):
+        pass
+
+
 @pytest.mark.parametrize('module', [
     # buit-in, no parameters
     nn.ReLU,
@@ -217,6 +226,7 @@ def test_no_warning_when_repeat(module):
 
 @pytest.mark.parametrize('module', [
     UserDefinedLayerWithParameters,
+    UserDefinedLayerWithBuffer,
 ])
 def test_warning_when_repeat(module):
     model = ppe.nn.ExtendedSequential(module())

--- a/tests/pytorch_pfn_extras_tests/nn_tests/modules_tests/test_extended_sequential.py
+++ b/tests/pytorch_pfn_extras_tests/nn_tests/modules_tests/test_extended_sequential.py
@@ -18,9 +18,11 @@ class UserDefinedLayer(nn.Module):
         pass
 
 
-@pytest.mark.parametrize('module', [nn.Sequential,
-                                    nn.ModuleList,
-                                    nn.ModuleDict])
+@pytest.mark.parametrize('container', [
+    nn.Sequential,
+    nn.ModuleList,
+    nn.ModuleDict,
+])
 @pytest.mark.parametrize('irregular_layer', [
     UserDefinedLayer,
     # No reset_parameters
@@ -33,26 +35,27 @@ class UserDefinedLayer(nn.Module):
         nn.MultiheadAttention, 1, 1),
     # ppe.nn layer
     functools.partial(
-        ppe.nn.LazyConv1d, None, 1, 1)])
+        ppe.nn.LazyConv1d, None, 1, 1),
+])
 class TestExtendedSequential(object):
 
     @pytest.fixture(autouse=True)
-    def setUp(self, module, irregular_layer):
+    def setUp(self, container, irregular_layer):
         self.l1 = ppe.nn.LazyLinear(None, 3)
         self.l2 = nn.Linear(3, 2)
         self.l3 = nn.Linear(2, 3)
         # a layer without reset_parameters
         self.l4 = irregular_layer()
         # s1: l1 -> l2
-        if module == nn.Sequential:
-            self.s1 = module(self.l1, self.l2)
-        elif module == nn.ModuleDict:
-            self.s1 = module({
+        if container == nn.Sequential:
+            self.s1 = container(self.l1, self.l2)
+        elif container == nn.ModuleDict:
+            self.s1 = container({
                 'l1': self.l1,
                 'l2': self.l2})
         else:
-            self.s1 = module([self.l1, self.l2])
-        self.module = module
+            self.s1 = container([self.l1, self.l2])
+        self.container = container
         # s2: s1 (l1 -> l2) -> l3 -> l4
         self.s2 = ppe.nn.ExtendedSequential(self.s1, self.l3, self.l4)
 
@@ -65,7 +68,7 @@ class TestExtendedSequential(object):
         assertions.assertIs(type(ret[1]), type(self.s2))
 
         # bias is filled with 0, so they should have the same values
-        if self.module == nn.ModuleDict:
+        if self.container == nn.ModuleDict:
             numpy.testing.assert_array_equal(
                 ret[0][0]['l1'].bias.detach().numpy(),
                 ret[1][0]['l1'].bias.detach().numpy())
@@ -104,7 +107,7 @@ class TestExtendedSequential(object):
         assertions.assertIsNot(ret[0], ret[1])
 
         # b is filled with 0, so they should have the same values
-        if self.module == nn.ModuleDict:
+        if self.container == nn.ModuleDict:
             numpy.testing.assert_array_equal(
                 ret[0][0]["l1"].bias.detach().numpy(),
                 ret[1][0]["l1"].bias.detach().numpy())
@@ -139,7 +142,7 @@ class TestExtendedSequential(object):
         assertions.assertIs(type(ret[1]), type(self.s2))
 
         # b is filled with 0, so they should have the same values
-        if self.module == nn.ModuleDict:
+        if self.container == nn.ModuleDict:
             numpy.testing.assert_array_equal(
                 ret[0][0]["l1"].bias.detach().numpy(),
                 ret[1][0]["l1"].bias.detach().numpy())

--- a/tests/pytorch_pfn_extras_tests/nn_tests/modules_tests/test_extended_sequential.py
+++ b/tests/pytorch_pfn_extras_tests/nn_tests/modules_tests/test_extended_sequential.py
@@ -16,8 +16,6 @@ class UserDefinedLayer(nn.Module):
     def forward(self):
         pass
 
-    def reset_parameters(self):
-        pass
 
 @pytest.mark.parametrize('module', [nn.Sequential,
                                     nn.ModuleList,

